File: sdks/python/src/mcparmor/testkit.py
"""MCP Armor TestKit — test your armor.json policies against the real broker.

Provides :class:`ArmorTestHarness`, a context manager that spins up the real
mcparmor broker backed by a lightweight mock MCP tool server. You define what
responses the mock tool returns; the harness sends ``tools/call`` messages
through the broker and reports whether they were blocked, allowed, or had
secrets redacted.

Example::

    async with ArmorTestHarness(armor="./armor.json") as harness:
        harness.mock_tool_response({
            "content": [{"type": "text", "text": "hello"}]
        })
        result = await harness.call_tool("read_file", {"path": "/etc/passwd"})
        assert result.blocked
        assert result.error_code == ArmorErrorCode.PATH_VIOLATION
"""
from __future__ import annotations

import asyncio
import json
import logging
import os
import subprocess
import sys
import tempfile
from dataclasses import dataclass
from pathlib import Path
from types import TracebackType
from typing import Any

from mcparmor._binary import find_binary

_logger = logging.getLogger(__name__)


class ArmorErrorCode:
    """JSON-RPC error codes returned by the mcparmor broker on policy violations."""

    PATH_VIOLATION = -32001
    NETWORK_VIOLATION = -32002
    SPAWN_VIOLATION = -32003
    SECRET_BLOCKED = -32004
    TIMEOUT = -32005


@dataclass
class ToolCallResult:
    """Outcome of sending a ``tools/call`` message through the broker.

    Attributes:
        raw: The full JSON-RPC response envelope.
        blocked: True if the broker returned an error (policy violation).
        allowed: True if the call passed through to the mock tool.
        error_code: The JSON-RPC error code, or None if not blocked.
        error_message: The error message string, or None if not blocked.
        response: The ``result`` payload from the mock tool, or None if blocked.
    """

    raw: dict[str, Any]
    blocked: bool
    allowed: bool
    error_code: int | None = None
    error_message: str | None = None
    response: dict[str, Any] | None = None

    @property
    def text(self) -> str | None:
        """Extract the first text content from the tool response.

        Returns:
            The text string if the response has MCP content items, None otherwise.
        """
        if self.response is None:
            return None
        content = self.response.get("content", [])
        if not content:
            return None
        return content[0].get("text")


class ArmorTestHarnessError(Exception):
    """Raised when the test harness cannot be started or communication fails."""


class ArmorTestHarness:
    """Test harness that runs the real mcparmor broker with a mock tool behind it.

    Use as an async context manager to manage the broker lifecycle::

        async with ArmorTestHarness(armor="./armor.json") as harness:
            result = await harness.call_tool("my_tool", {"key": "value"})

    Args:
        armor: Path to the armor.json manifest to test.
        profile: Optional profile override (e.g. ``"strict"``).
        no_os_sandbox: Disable Layer 2 OS sandbox. Defaults to True because
            testkit tests Layer 1 enforcement; the OS sandbox would interfere
            with the mock server.
        timeout: Read timeout in seconds for individual tool calls.
    """

    def __init__(
        self,
        *,
        armor: str | Path,
        profile: str | None = None,
        no_os_sandbox: bool = True,
        timeout: float = 10.0,
    ) -> None:
        self._armor = str(armor)
        self._profile = profile
        self._no_os_sandbox = no_os_sandbox
        self._timeout = timeout
        self._process: subprocess.Popen[bytes] | None = None
        self._tmp_dir: tempfile.TemporaryDirectory[str] | None = None
        self._config_path: str = ""
        self._next_id_counter = 0
        self._config: dict[str, Any] = {
            "server_info": {"name": "mcparmor-mock-tool", "version": "1.0"},
            "tools": [],
            "default_response": {
                "content": [{"type": "text", "text": "mock response"}],
            },
            "responses": {},
        }

    # ------------------------------------------------------------------
    # Context manager
    # ------------------------------------------------------------------

    async def __aenter__(self) -> ArmorTestHarness:
        """Start the broker process and perform the MCP handshake."""
        self._start()
        await self._handshake()
        return self

    async def __aexit__(
        self,
        exc_type: type[BaseException] | None,
        exc_val: BaseException | None,
        exc_tb: TracebackType | None,
    ) -> None:
        """Terminate the broker process and clean up temp files."""
        self.stop()

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def mock_tool_response(
        self,
        response: dict[str, Any],
        *,
        tool_name: str | None = None,
    ) -> None:
        """Set the response the mock tool returns for ``tools/call``.

        Args:
            response: The MCP result payload (e.g.
                ``{"content": [{"type": "text", "text": "..."}]}``).
            tool_name: If given, this response is used only when the tool call
                targets this specific tool name. Otherwise, it becomes the
                default response for all tools.
        """
        if tool_name is not None:
            self._config["responses"][tool_name] = response
        else:
            self._config["default_response"] = response
        self._write_config()

    def set_tools(self, tools: list[dict[str, Any]]) -> None:
        """Set the tool definitions returned by ``tools/list``.

        Args:
            tools: List of MCP tool definition objects (name, description,
                inputSchema).
        """
        self._config["tools"] = tools
        self._write_config()

    async def call_tool(
        self,
        name: str,
        arguments: dict[str, Any] | None = None,
    ) -> ToolCallResult:
        """Send a ``tools/call`` JSON-RPC message through the broker.

        Args:
            name: The tool name to call.
            arguments: The arguments to pass to the tool.

        Returns:
            A :class:`ToolCallResult` describing whether the call was blocked
            or allowed, and the response payload.

        Raises:
            ArmorTestHarnessError: If the broker process is not running.
        """
        if arguments is None:
            arguments = {}

        message = {
            "jsonrpc": "2.0",
            "id": self._next_id(),
            "method": "tools/call",
            "params": {
                "name": name,
                "arguments": arguments,
            },
        }
        raw_response = await self._send_and_receive(message)
        return _classify_response(raw_response)

    async def send_raw(self, message: dict[str, Any]) -> dict[str, Any]:
        """Send an arbitrary JSON-RPC message and return the raw response.

        Useful for testing non-``tools/call`` interactions (e.g. ``tools/list``).

        Args:
            message: A JSON-RPC request object. Must include ``id`` if a
                response is expected.

        Returns:
            The parsed JSON-RPC response dictionary.
        """
        return await self._send_and_receive(message)

    def stop(self) -> None:
        """Terminate the broker and clean up resources.

        Safe to call multiple times.
        """
        if self._process is not None:
            try:
                self._process.terminate()
                self._process.wait(timeout=5)
            except OSError:
                _logger.warning("Failed to terminate broker cleanly")
            finally:
                self._process = None
        if self._tmp_dir is not None:
            self._tmp_dir.cleanup()
            self._tmp_dir = None

    # ------------------------------------------------------------------
    # Private helpers
    # ------------------------------------------------------------------

    def _next_id(self) -> int:
        """Return an auto-incrementing JSON-RPC message ID."""
        self._next_id_counter += 1
        return self._next_id_counter

    def _write_config(self) -> None:
        """Persist the current mock config to disk atomically.

        Writes to a temp file first then renames, so the mock server never
        sees a truncated/empty config file.
        """
        tmp_path = self._config_path + ".tmp"
        with open(tmp_path, "w") as fh:
            json.dump(self._config, fh)
        os.replace(tmp_path, self._config_path)

    def _start(self) -> None:
        """Start the broker with the mock tool server behind it."""
        self._tmp_dir = tempfile.TemporaryDirectory(prefix="mcparmor-testkit-")
        self._config_path = os.path.join(self._tmp_dir.name, "mock_config.json")
        self._write_config()

        binary = str(find_binary())
        mock_server = str(
            Path(__file__).parent / "_mock_tool_server.py"
        )

        broker_args = ["run"]
        broker_args.extend(["--armor", self._armor])
        if self._profile is not None:
            broker_args.extend(["--profile", self._profile])
        if self._no_os_sandbox:
            broker_args.append("--no-os-sandbox")
        broker_args.extend(["--no-audit", "--"])
        broker_args.extend([sys.executable, mock_server, self._config_path])

        env = os.environ.copy()

        try:
            self._process = subprocess.Popen(
                [binary, *broker_args],
                stdin=subprocess.PIPE,
                stdout=subprocess.PIPE,
                stderr=subprocess.DEVNULL,
                env=env,
            )
        except OSError as exc:
            raise ArmorTestHarnessError(
                f"Failed to start mcparmor broker: {exc}"
            ) from exc

    async def _handshake(self) -> None:
        """Perform the MCP initialize / notifications/initialized handshake."""
        init_msg = {
            "jsonrpc": "2.0",
            "id": self._next_id(),
            "method": "initialize",
            "params": {
                "protocolVersion": "2024-11-05",
                "capabilities": {},
                "clientInfo": {"name": "mcparmor-testkit", "version": "1.0"},
            },
        }
        response = await self._send_and_receive(init_msg)
        if "error" in response:
            raise ArmorTestHarnessError(
                f"MCP handshake failed: {response['error']}"
            )

        proc = self._process
        if proc is None or proc.stdin is None:
            raise ArmorTestHarnessError("Broker process is not running.")
        notification = json.dumps({
            "jsonrpc": "2.0",
            "method": "notifications/initialized",
        }) + "\n"
        proc.stdin.write(notification.encode())
        proc.stdin.flush()

    async def _send_and_receive(
        self, message: dict[str, Any]
    ) -> dict[str, Any]:
        """Write a message and read one JSON-RPC response line.

        Performs both the write and blocking read in a thread to avoid
        blocking the event loop and to ensure atomicity.

        Args:
            message: JSON-RPC request to send.

        Returns:
            Parsed JSON-RPC response.

        Raises:
            ArmorTestHarnessError: If the broker is not running or the read
                times out.
        """
        loop = asyncio.get_running_loop()
        try:
            return await asyncio.wait_for(
                loop.run_in_executor(
                    None, self._sync_send_and_receive, message
                ),
                timeout=self._timeout,
            )
        except asyncio.TimeoutError:
            raise ArmorTestHarnessError(
                "Timed out waiting for broker response."
            )

    def _sync_send_and_receive(
        self, message: dict[str, Any]
    ) -> dict[str, Any]:
        """Write a message and read one response, synchronously.

        Args:
            message: JSON-RPC request to send.

        Returns:
            Parsed JSON-RPC response dict.
        """
        proc = self._process
        if proc is None or proc.stdin is None or proc.stdout is None:
            raise ArmorTestHarnessError("Broker process is not running.")

        line = json.dumps(message) + "\n"
        proc.stdin.write(line.encode())
        proc.stdin.flush()

        raw = proc.stdout.readline()
        if not raw:
            raise ArmorTestHarnessError(
                "Broker closed stdout without responding."
            )

        text = raw.decode().strip()
        try:
            return json.loads(text)
        except json.JSONDecodeError as exc:
            raise ArmorTestHarnessError(
                f"Broker returned invalid JSON: {text}"
            ) from exc


def _classify_response(raw: dict[str, Any]) -> ToolCallResult:
    """Classify a JSON-RPC response as blocked or allowed.

    Args:
        raw: The full JSON-RPC response envelope.

    Returns:
        A populated :class:`ToolCallResult`.
    """
    if "error" in raw:
        error = raw["error"]
        return ToolCallResult(
            raw=raw,
            blocked=True,
            allowed=False,
            error_code=error.get("code"),
            error_message=error.get("message"),
        )
    return ToolCallResult(
        raw=raw,
        blocked=False,
        allowed=True,
        response=raw.get("result"),
    )

File: sdks/node/src/testkit.ts
/**
 * MCP Armor TestKit — test your armor.json policies against the real broker.
 *
 * Provides {@link ArmorTestHarness}, which spins up the real mcparmor broker
 * backed by a lightweight mock MCP tool server. You define what responses the
 * mock tool returns; the harness sends `tools/call` messages through the
 * broker and reports whether they were blocked, allowed, or had secrets
 * redacted.
 *
 * @example
 * ```typescript
 * const harness = await ArmorTestHarness.start({ armor: './armor.json' });
 * harness.mockToolResponse({ content: [{ type: 'text', text: 'hello' }] });
 * const result = await harness.callTool('read_file', { path: '/etc/passwd' });
 * assert(result.blocked);
 * await harness.stop();
 * ```
 */

import { spawn, type ChildProcess } from 'node:child_process';
import { writeFileSync, mkdtempSync, rmSync, renameSync } from 'node:fs';
import { tmpdir } from 'node:os';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';
import { findBinary } from './binary.ts';

// ---------------------------------------------------------------------------
// Public types
// ---------------------------------------------------------------------------

/** JSON-RPC error codes returned by the mcparmor broker on policy violations. */
export const ArmorErrorCode = {
  PATH_VIOLATION: -32001,
  NETWORK_VIOLATION: -32002,
  SPAWN_VIOLATION: -32003,
  SECRET_BLOCKED: -32004,
  TIMEOUT: -32005,
} as const;

/** The outcome of sending a `tools/call` message through the broker. */
export interface ToolCallResult {
  /** The full JSON-RPC response envelope. */
  readonly raw: Record<string, unknown>;
  /** True if the broker returned an error (policy violation). */
  readonly blocked: boolean;
  /** True if the call passed through to the mock tool. */
  readonly allowed: boolean;
  /** The JSON-RPC error code, or null if not blocked. */
  readonly errorCode: number | null;
  /** The error message string, or null if not blocked. */
  readonly errorMessage: string | null;
  /** The `result` payload from the mock tool, or null if blocked. */
  readonly response: Record<string, unknown> | null;
  /** The first text content from the tool response, or null. */
  readonly text: string | null;
}

/** Options for creating an {@link ArmorTestHarness}. */
export interface ArmorTestHarnessOptions {
  /** Path to the armor.json manifest to test. */
  readonly armor: string;
  /** Optional profile override (e.g. `"strict"`). */
  readonly profile?: string;
  /**
   * Disable Layer 2 OS sandbox. Defaults to `true` because the testkit tests
   * Layer 1 enforcement; the OS sandbox would interfere with the mock server.
   */
  readonly noOsSandbox?: boolean;
  /** Read timeout in milliseconds for individual tool calls. Default: 10 000. */
  readonly timeoutMs?: number;
}

// ---------------------------------------------------------------------------
// Error type
// ---------------------------------------------------------------------------

/** Thrown when the test harness cannot be started or communication fails. */
export class ArmorTestHarnessError extends Error {
  constructor(message: string, options?: { cause?: unknown }) {
    super(message, options);
    this.name = 'ArmorTestHarnessError';
  }
}

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

const DEFAULT_TIMEOUT_MS = 10_000;

/** Resolve the path to the bundled mock-tool-server.js. */
function resolveMockServerPath(): string {
  const currentDir = dirname(fileURLToPath(import.meta.url));
  return join(currentDir, 'mock-tool-server.js');
}

/**
 * Read exactly one newline-terminated line from the process stdout.
 *
 * @param proc - The child process to read from.
 * @param timeoutMs - Maximum milliseconds to wait.
 * @returns Parsed JSON response object.
 */
function readLine(
  proc: ChildProcess,
  timeoutMs: number,
): Promise<Record<string, unknown>> {
  return new Promise((resolve, reject) => {
    if (proc.stdout === null) {
      reject(new ArmorTestHarnessError('Process stdout is not available'));
      return;
    }

    let buffer = '';
    let settled = false;

    const timer = setTimeout(() => {
      if (settled) return;
      settled = true;
      cleanup();
      reject(new ArmorTestHarnessError('Timed out waiting for broker response'));
    }, timeoutMs);

    const onData = (chunk: Buffer | string): void => {
      if (settled) return;
      buffer += chunk.toString();
      const idx = buffer.indexOf('\n');
      if (idx === -1) return;

      const line = buffer.slice(0, idx);
      settled = true;
      cleanup();

      try {
        const parsed = JSON.parse(line) as Record<string, unknown>;
        resolve(parsed);
      } catch (err) {
        reject(new ArmorTestHarnessError(`Broker returned invalid JSON: ${line}`, { cause: err }));
      }
    };

    const onClose = (): void => {
      if (settled) return;
      settled = true;
      cleanup();
      reject(new ArmorTestHarnessError('Broker closed stdout without responding'));
    };

    const cleanup = (): void => {
      clearTimeout(timer);
      proc.stdout?.removeListener('data', onData);
      proc.stdout?.removeListener('close', onClose);
    };

    proc.stdout.on('data', onData);
    proc.stdout.once('close', onClose);
  });
}

/**
 * Write a JSON-RPC message to the process stdin.
 *
 * @param proc - The child process whose stdin to write to.
 * @param message - The message to serialize and send.
 */
function writeMessage(proc: ChildProcess, message: Record<string, unknown>): Promise<void> {
  return new Promise((resolve, reject) => {
    if (proc.stdin === null) {
      reject(new ArmorTestHarnessError('Process stdin is not available'));
      return;
    }
    const line = JSON.stringify(message) + '\n';
    const onDrain = (): void => resolve();
    const flushed = proc.stdin.write(line, (err) => {
      if (err != null) {
        proc.stdin?.removeListener('drain', onDrain);
        reject(new ArmorTestHarnessError('Failed to write to broker stdin', { cause: err }));
      }
    });
    if (flushed) {
      resolve();
    } else {
      proc.stdin.once('drain', onDrain);
    }
  });
}

/**
 * Classify a JSON-RPC response as blocked or allowed.
 *
 * @param raw - The full JSON-RPC response envelope.
 * @returns A populated {@link ToolCallResult}.
 */
function classifyResponse(raw: Record<string, unknown>): ToolCallResult {
  const error = raw['error'] as Record<string, unknown> | undefined;
  if (error !== undefined) {
    return {
      raw,
      blocked: true,
      allowed: false,
      errorCode: (error['code'] as number) ?? null,
      errorMessage: (error['message'] as string) ?? null,
      response: null,
      text: null,
    };
  }

  const response = (raw['result'] as Record<string, unknown>) ?? null;
  let text: string | null = null;
  if (response !== null) {
    const content = response['content'] as Array<Record<string, unknown>> | undefined;
    if (content !== undefined && content.length > 0) {
      text = (content[0]?.['text'] as string) ?? null;
    }
  }

  return {
    raw,
    blocked: false,
    allowed: true,
    errorCode: null,
    errorMessage: null,
    response,
    text,
  };
}

// ---------------------------------------------------------------------------
// ArmorTestHarness
// ---------------------------------------------------------------------------

/**
 * Test harness that runs the real mcparmor broker with a mock tool behind it.
 *
 * Use the static {@link ArmorTestHarness.start} factory to create and
 * initialise an instance. Call {@link ArmorTestHarness.stop} when done.
 */
export class ArmorTestHarness {
  readonly #armor: string;
  readonly #profile: string | undefined;
  readonly #noOsSandbox: boolean;
  readonly #timeoutMs: number;
  #process: ChildProcess | null = null;
  #tmpDir: string | null = null;
  #configPath: string = '';
  #idCounter: number = 0;
  #config: Record<string, unknown>;

  private constructor(options: ArmorTestHarnessOptions) {
    this.#armor = options.armor;
    this.#profile = options.profile;
    this.#noOsSandbox = options.noOsSandbox ?? true;
    this.#timeoutMs = options.timeoutMs ?? DEFAULT_TIMEOUT_MS;
    this.#config = {
      server_info: { name: 'mcparmor-mock-tool', version: '1.0' },
      tools: [],
      default_response: {
        content: [{ type: 'text', text: 'mock response' }],
      },
      responses: {},
    };
  }

  /**
   * Create and start a test harness.
   *
   * Spawns the real mcparmor broker, performs the MCP handshake, and returns
   * a ready-to-use harness.
   *
   * @param options - Configuration for the test harness.
   * @returns A running {@link ArmorTestHarness}.
   */
  static async start(options: ArmorTestHarnessOptions): Promise<ArmorTestHarness> {
    const harness = new ArmorTestHarness(options);
    harness.#spawn();
    await harness.#handshake();
    return harness;
  }

  // ------------------------------------------------------------------
  // Public API
  // ------------------------------------------------------------------

  /**
   * Set the response the mock tool returns for `tools/call`.
   *
   * @param response - The MCP result payload.
   * @param toolName - If given, this response is used only for this tool name.
   *   Otherwise it becomes the default response for all tools.
   */
  mockToolResponse(response: Record<string, unknown>, toolName?: string): void {
    if (toolName !== undefined) {
      (this.#config['responses'] as Record<string, unknown>)[toolName] = response;
    } else {
      this.#config['default_response'] = response;
    }
    this.#writeConfig();
  }

  /**
   * Set the tool definitions returned by `tools/list`.
   *
   * @param tools - List of MCP tool definition objects.
   */
  setTools(tools: ReadonlyArray<Record<string, unknown>>): void {
    this.#config['tools'] = [...tools];
    this.#writeConfig();
  }

  /**
   * Send a `tools/call` JSON-RPC message through the broker.
   *
   * @param name - The tool name to call.
   * @param args - The arguments to pass to the tool.
   * @returns A {@link ToolCallResult} describing the outcome.
   */
  async callTool(
    name: string,
    args: Record<string, unknown> = {},
  ): Promise<ToolCallResult> {
    const message = {
      jsonrpc: '2.0',
      id: this.#nextId(),
      method: 'tools/call',
      params: { name, arguments: args },
    };
    const raw = await this.#sendAndReceive(message);
    return classifyResponse(raw);
  }

  /**
   * Send an arbitrary JSON-RPC message and return the raw response.
   *
   * @param message - A JSON-RPC request object.
   * @returns The parsed JSON-RPC response.
   */
  async sendRaw(message: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.#sendAndReceive(message);
  }

  /**
   * Terminate the broker and clean up resources.
   *
   * @returns A promise that resolves when the process has exited.
   */
  async stop(): Promise<void> {
    const proc = this.#process;
    if (proc !== null) {
      this.#process = null;
      await new Promise<void>((resolve) => {
        proc.once('close', () => resolve());
        proc.kill();
      });
    }
    if (this.#tmpDir !== null) {
      rmSync(this.#tmpDir, { recursive: true, force: true });
      this.#tmpDir = null;
    }
  }

  // ------------------------------------------------------------------
  // Private helpers
  // ------------------------------------------------------------------

  #nextId(): number {
    this.#idCounter += 1;
    return this.#idCounter;
  }

  /** Write the config atomically (write to .tmp then rename). */
  #writeConfig(): void {
    const tmpPath = this.#configPath + '.tmp';
    writeFileSync(tmpPath, JSON.stringify(this.#config));
    renameSync(tmpPath, this.#configPath);
  }

  #spawn(): void {
    this.#tmpDir = mkdtempSync(join(tmpdir(), 'mcparmor-testkit-'));
    this.#configPath = join(this.#tmpDir, 'mock_config.json');
    this.#writeConfig();

    const binaryPath = findBinary();
    const mockServerPath = resolveMockServerPath();

    const brokerArgs: string[] = ['run', '--armor', this.#armor];
    if (this.#profile !== undefined) {
      brokerArgs.push('--profile', this.#profile);
    }
    if (this.#noOsSandbox) {
      brokerArgs.push('--no-os-sandbox');
    }
    brokerArgs.push('--no-audit', '--', process.execPath, mockServerPath, this.#configPath);

    this.#process = spawn(binaryPath, brokerArgs, {
      stdio: 'pipe',
    });
  }

  async #handshake(): Promise<void> {
    const initMsg = {
      jsonrpc: '2.0',
      id: this.#nextId(),
      method: 'initialize',
      params: {
        protocolVersion: '2024-11-05',
        capabilities: {},
        clientInfo: { name: 'mcparmor-testkit', version: '1.0' },
      },
    };
    const response = await this.#sendAndReceive(initMsg);
    if ('error' in response) {
      throw new ArmorTestHarnessError(
        `MCP handshake failed: ${JSON.stringify(response['error'])}`,
      );
    }

    const notification = {
      jsonrpc: '2.0',
      method: 'notifications/initialized',
    };
    await writeMessage(this.#process!, notification);
  }

  async #sendAndReceive(
    message: Record<string, unknown>,
  ): Promise<Record<string, unknown>> {
    const proc = this.#process;
    if (proc === null) {
      throw new ArmorTestHarnessError('Broker process is not running');
    }
    await writeMessage(proc, message);
    return readLine(proc, this.#timeoutMs);
  }
}
